File: src/leetcode/editor/cn/P51NQueens.java
//n 皇后问题研究的是如何将 n 个皇后放置在 n×n 的棋盘上，并且使皇后彼此之间不能相互攻击。 
//
// 
//
// 上图为 8 皇后问题的一种解法。 
//
// 给定一个整数 n，返回所有不同的 n 皇后问题的解决方案。 
//
// 每一种解法包含一个明确的 n 皇后问题的棋子放置方案，该方案中 'Q' 和 '.' 分别代表了皇后和空位。 
//
// 示例: 
//
// 输入: 4
//输出: [
// [".Q..",  // 解法 1
//  "...Q",
//  "Q...",
//  "..Q."],
//
// ["..Q.",  // 解法 2
//  "Q...",
//  "...Q",
//  ".Q.."]
//]
//解释: 4 皇后问题存在两个不同的解法。
// 
//
// 
//
// 提示： 
//
// 
// 皇后，是国际象棋中的棋子，意味着国王的妻子。皇后只做一件事，那就是“吃子”。当她遇见可以吃的棋子时，就迅速冲上去吃掉棋子。当然，她横、竖、斜都可走一到七步
//，可进可退。（引用自 百度百科 - 皇后 ） 
// 
// Related Topics 回溯算法


package leetcode.editor.cn;

import java.util.*;

//Java：N皇后
public class P51NQueens{
    public static void main(String[] args) {
        Solution solution = new P51NQueens().new Solution();
        // TO TEST
        System.out.println(solution.solveNQueens(4));
    }
    //leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    List<List<String>> res= new ArrayList<>();
    public List<List<String>> solveNQueens(int n) {
        int[] queens= new int[n];
        Arrays.fill(queens,-1);
        backtrack(queens, n, 0, new HashSet<>(), new HashSet<>(), new HashSet<>());
        return res;
    }

    public void backtrack(int[] queens, int n, int row,
        Set<Integer> columns, Set<Integer> diagonals1, Set<Integer> diagonals2){
        if(row==n){
            List<String> ls= generate(queens, n);
            res.add(ls);
            return;
        }
        for(int i=0;i<n;i++){
            if(columns.contains(i))
                continue;
            if(diagonals1.contains(row-i))
                continue;
            if(diagonals2.contains(row+i))
                continue;
            queens[row]= i;
            columns.add(i);
            diagonals1.add(row-i);
            diagonals2.add(row+i);
            backtrack(queens, n,row+1, columns, diagonals1, diagonals2);
            queens[row]= -1;
            columns.remove(i);
            diagonals1.remove(row-i);
            diagonals2.remove(row+i);
        }
    }

    public List<String> generate(int[] queens, int n){
        List<String> board= new ArrayList<>();
        for(int i=0;i<queens.length;i++){
            char[] row= new char[n];
            Arrays.fill(row, '.');
            row[queens[i]]= 'Q';
            board.add(new String(row));
        }
        return board;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

}

File: leetgo/leetcode/editor/cn/P51NQueens.go
//n 皇后问题研究的是如何将 n 个皇后放置在 n×n 的棋盘上，并且使皇后彼此之间不能相互攻击。
//
//
//
// 上图为 8 皇后问题的一种解法。
//
// 给定一个整数 n，返回所有不同的 n 皇后问题的解决方案。
//
// 每一种解法包含一个明确的 n 皇后问题的棋子放置方案，该方案中 'Q' 和 '.' 分别代表了皇后和空位。
//
//
//
// 示例：
//
// 输入：4
//输出：[
// [".Q..",  // 解法 1
//  "...Q",
//  "Q...",
//  "..Q."],
//
// ["..Q.",  // 解法 2
//  "Q...",
//  "...Q",
//  ".Q.."]
//]
//解释: 4 皇后问题存在两个不同的解法。
//
//
//
//
// 提示：
//
//
// 皇后彼此不能相互攻击，也就是说：任何两个皇后都不能处于同一条横行、纵行或斜线上。
//
// Related Topics 回溯算法

package main

//Go：N 皇后
import "fmt"

func main() {
	fmt.Println(solveNQueens(4))
}

//leetcode submit region begin(Prohibit modification and deletion)
var solutions [][]string

func solveNQueens(n int) [][]string {
	solutions = [][]string{}
	queen := make([]int, n)
	for i := 0; i < n; i++ {
		queen[i] = -1
	}
	column := map[int]bool{}
	diagonals1, diagonals2 := map[int]bool{}, map[int]bool{}
	backtrack(queen, n, 0, column, diagonals1, diagonals2)
	return solutions
}

func backtrack(queen []int, n, row int, column, diagonals1, diagonals2 map[int]bool) {
	if row == n {
		board := generateBoard(queen, n)
		solutions = append(solutions, board)
		return
	}

	for i := 0; i < n; i++ {
		if column[i] {
			continue
		}
		//foreach diagonals
		if diagonals1[row-i] {
			continue
		}
		if diagonals2[row+i] {
			continue
		}
		queen[row] = i
		column[i] = true
		diagonals1[row-i], diagonals2[row+i] = true, true
		backtrack(queen, n, row+1, column, diagonals1, diagonals2)
		queen[row] = -1
		delete(column, i)
		delete(diagonals1, row-i)
		delete(diagonals2, row+i)
	}
}

func generateBoard(queen []int, n int) []string {
	board := []string{}
	for i := 0; i < n; i++ {
		row := make([]byte, n)
		for j := 0; j < n; j++ {
			row[j] = '.'
		}
		row[queen[i]] = 'Q'
		board = append(board, string(row))
	}
	return board
}

//leetcode submit region end(Prohibit modification and deletion)
